File: python/collect_ist_gateway_b.py
#!/usr/bin/env python3
# collect_ist_gateway_b.sh — portiert nach python
# Quelle: shell, OpenClaw@gateway2:scripts/collect_ist_gateway_b.sh
# Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

import os
import sys
import subprocess
import json
import shutil
from datetime import datetime
from pathlib import Path
import urllib.request
import urllib.error

def run_command(command, shell=False, capture_output=True, text=True):
    """Helper function to run shell commands"""
    try:
        if shell:
            result = subprocess.run(command, shell=True, capture_output=capture_output, text=text)
        else:
            result = subprocess.run(command, capture_output=capture_output, text=text)
        return result.stdout.strip() if capture_output else ""
    except (subprocess.CalledProcessError, FileNotFoundError):
        return ""

def get_public_ip():
    """Get public IP address"""
    try:
        with urllib.request.urlopen('https://ifconfig.me', timeout=4) as response:
            return response.read().decode('utf-8').strip()
    except Exception:
        return "(nicht ermittelt)"

def get_tailscale_ip():
    """Get Tailscale IP address"""
    try:
        result = subprocess.run(['tailscale', 'ip', '-4'], capture_output=True, text=True, timeout=10)
        if result.returncode == 0:
            return result.stdout.strip().split('\n')[0]
        else:
            return "(nicht ermittelt)"
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return "(nicht ermittelt)"

def get_os_pretty_name():
    """Get OS pretty name from /etc/os-release"""
    try:
        with open('/etc/os-release', 'r') as f:
            for line in f:
                if line.startswith('PRETTY_NAME='):
                    return line.split('=', 1)[1].strip().strip('"')
    except FileNotFoundError:
        pass
    return ""

def check_file_exists(filepath):
    """Check if file exists and return appropriate string"""
    return "vorhanden" if os.path.exists(filepath) else "fehlt"

def main():
    # Set variables
    BASE_DIR = os.path.expanduser("~/.openclaw")
    OUT_DIR = os.path.join(BASE_DIR, "workspace/vscode")
    NOW_UTC = datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ")
    NOW_LOCAL = datetime.now().strftime("%Y-%m-%d %H:%M:%S %Z")
    TS = datetime.now().strftime("%Y%m%d-%H%M%S")

    # Create output directory
    os.makedirs(OUT_DIR, exist_ok=True)

    # File paths
    IST_FILE = os.path.join(OUT_DIR, "IST-ZUSTAND_GATEWAY-B_NODE7.md")
    INV_FILE = os.path.join(OUT_DIR, "ARTEFAKT-INVENTAR_GATEWAY-B_NODE7.md")
    CFG_FILE = os.path.join(OUT_DIR, "OPENCLAW-CONFIG-SNAPSHOT_GATEWAY-B_NODE7.md")
    ENV_FILE = os.path.join(OUT_DIR, "ENV-STATUS_GATEWAY-B_NODE7.md")
    RUN_FILE = os.path.join(OUT_DIR, f"RUN-{TS}.md")

    OPENCLAW_JSON = os.path.join(BASE_DIR, "openclaw.json")
    ENV_DOT = os.path.join(BASE_DIR, ".env")
    ENV_SYSTEMD = os.path.join(BASE_DIR, "gateway.systemd.env")
    VSCODE_DIR = os.path.join(BASE_DIR, ".vscode")

    # System information
    HOSTNAME_FQDN = run_command(["hostname", "-f"]) or run_command(["hostname"])
    HOSTNAME_SHORT = run_command(["hostname"])
    ARCH = run_command(["uname", "-m"])
    KERNEL = run_command(["uname", "-r"])
    OS_PRETTY = get_os_pretty_name()
    IPV4_ALL = run_command(["hostname", "-I"]).strip()
    PUBLIC_IP = get_public_ip()
    TAILSCALE_IP = get_tailscale_ip()
    
    try:
        OPENCLAW_VER = run_command(["openclaw", "--version"])
    except FileNotFoundError:
        OPENCLAW_VER = "(nicht ermittelt)"
        
    try:
        NODE_VER = run_command(["node", "-v"])
    except FileNotFoundError:
        NODE_VER = "(nicht ermittelt)"

    # Fill in defaults if empty
    if not PUBLIC_IP:
        PUBLIC_IP = "(nicht ermittelt)"
    if not TAILSCALE_IP:
        TAILSCALE_IP = "(nicht ermittelt)"
    if not OPENCLAW_VER:
        OPENCLAW_VER = "(nicht ermittelt)"
    if not NODE_VER:
        NODE_VER = "(nicht ermittelt)"

    # Write IST file
    with open(IST_FILE, 'w') as f:
        f.write(f"""# IST-Zustand: Gateway B / Node 7

Stand (lokal): {NOW_LOCAL}  
Stand (UTC): {NOW_UTC}

## 1) Identität & System

- Gateway: **B**
- Node: **7**
- Hostname (short): `{HOSTNAME_SHORT}`
- Hostname (FQDN): `{HOSTNAME_FQDN}`
- Architektur: `{ARCH}`
- Kernel: `{KERNEL}`
- OS: `{OS_PRETTY}`
- IPv4 (lokal): `{IPV4_ALL}`
- Public IPv4: `{PUBLIC_IP}`
- Tailscale IPv4: `{TAILSCALE_IP}`
- OpenClaw Version: `{OPENCLAW_VER}`
- Node.js Version: `{NODE_VER}`

## 2) Arbeitsverzeichnisse

- Basis: `{BASE_DIR}`
- Funktionell VSCode: `{VSCODE_DIR}`
- Workspace Doku: `{OUT_DIR}`

## 3) Kernartefakte (Existenz)

- `{OPENCLAW_JSON}`: {check_file_exists(OPENCLAW_JSON)}
- `{ENV_DOT}`: {check_file_exists(ENV_DOT)}
- `{ENV_SYSTEMD}`: {check_file_exists(ENV_SYSTEMD)}
- `{os.path.join(BASE_DIR, "plugins/installs.json")}`: {check_file_exists(os.path.join(BASE_DIR, "plugins/installs.json"))}
- `{os.path.join(BASE_DIR, "plugin-skills")}`: {check_file_exists(os.path.join(BASE_DIR, "plugin-skills"))}

## 4) Hinweis

Diese Datei wird bei jedem Lauf neu geschrieben.
Zusätzlich wird ein Laufprotokoll als `RUN-*.md` erzeugt.
""")

    # Write inventory file
    with open(INV_FILE, 'w') as f:
        f.write(f"""# Artefakt-Inventar: Gateway B / Node 7

Stand: {NOW_LOCAL}

## Top-Level in ~/.openclaw

```
""")
        try:
            for item in sorted(os.listdir(BASE_DIR)):
                f.write(f"{item}\n")
        except FileNotFoundError:
            pass
        f.write("""```

## ~/.openclaw/.vscode

```
""")
        if os.path.exists(VSCODE_DIR):
            try:
                for item in sorted(os.listdir(VSCODE_DIR)):
                    item_path = os.path.join(VSCODE_DIR, item)
                    stat = os.stat(item_path)
                    f.write(f"{'d' if os.path.isdir(item_path) else '-'} {stat.st_size:>8} {item}\n")
            except FileNotFoundError:
                f.write("(nicht vorhanden)")
        else:
            f.write("(nicht vorhanden)")
        f.write("""```

## plugin-skills/

```
""")
        plugin_skills_dir = os.path.join(BASE_DIR, "plugin-skills")
        if os.path.exists(plugin_skills_dir):
            try:
                for item in sorted(os.listdir(plugin_skills_dir)):
                    f.write(f"{item}\n")
            except FileNotFoundError:
                f.write("(nicht vorhanden)")
        else:
            f.write("(nicht vorhanden)")
        f.write("""```

## openclaw.json Backups

```
""")
        try:
            backup_files = [f for f in os.listdir(BASE_DIR) if f.startswith("openclaw.json.bak")]
            if backup_files:
                for bf in sorted(backup_files):
                    f.write(f"{bf}\n")
            else:
                f.write("(keine gefunden)")
        except FileNotFoundError:
            f.write("(keine gefunden)")
        f.write("```\n")

    # Write config snapshot file
    with open(CFG_FILE, 'w') as f:
        f.write(f"""# OpenClaw Config Snapshot: Gateway B / Node 7

Stand: {NOW_LOCAL}

## Schlüsselpositionen (grep)

```
""")
        if os.path.exists(OPENCLAW_JSON):
            try:
                with open(OPENCLAW_JSON, 'r') as json_file:
                    lines = json_file.readlines()
                    for i, line in enumerate(lines, 1):
                        if any(key in line for key in ['"gateway"', '"session"', '"dmScope"', '"auth"', '"secrets"', '"tools"', '"plugins"', '"profile"', '"alsoAllow"', '"denyCommands"']):
                            f.write(f"{i}: {line}")
            except Exception:
                pass
        else:
            f.write("openclaw.json fehlt")
        f.write("""```

## Ausschnitt gateway/session/auth (ungefiltert, betriebsnah)

```json
""")
        if os.path.exists(OPENCLAW_JSON):
            try:
                with open(OPENCLAW_JSON, 'r') as json_file:
                    lines = json_file.readlines()
                    for line in lines[579:780]:  # 580-780 lines (0-indexed)
                        f.write(line)
            except Exception:
                f.write("{ \"error\": \"Fehler beim Lesen der Datei\" }\n")
        else:
            f.write("{ \"error\": \"openclaw.json fehlt\" }\n")
        f.write("```\n")

    # Write ENV status file
    with open(ENV_FILE, 'w') as f:
        f.write(f"""# ENV-Status: Gateway B / Node 7

Stand: {NOW_LOCAL}

## Dateien

```
""")
        env_files = [ENV_DOT, ENV_SYSTEMD]
        for ef in env_files:
            if os.path.exists(ef):
                stat = os.stat(ef)
                f.write(f"{'d' if os.path.isdir(ef) else '-'} {stat.st_size:>8} {ef}\n")
        f.write("""```

## .env (vollständig, ungefiltert)

```dotenv
""")
        if os.path.exists(ENV_DOT):
            try:
                with open(ENV_DOT, 'r') as env_file:
                    f.write(env_file.read())
            except Exception:
                f.write("# Fehler beim Lesen der Datei")
        else:
            f.write("# .env fehlt")
        f.write("""```

## gateway.systemd.env (vollständig, ungefiltert)

```dotenv
""")
        if os.path.exists(ENV_SYSTEMD):
            try:
                with open(ENV_SYSTEMD, 'r') as env_file:
                    f.write(env_file.read())
            except Exception:
                f.write("# Fehler beim Lesen der Datei")
        else:
            f.write("# gateway.systemd.env fehlt")
        f.write("```\n")

    # Write run file
    script_path = os.path.realpath(__file__)
    with open(RUN_FILE, 'w') as f:
        f.write(f"""# Laufprotokoll Gateway B / Node 7

- Zeit (lokal): {NOW_LOCAL}
- Zeit (UTC): {NOW_UTC}
- Script: {script_path}

## Erzeugte Dateien

- {os.path.basename(IST_FILE)}
- {os.path.basename(INV_FILE)}
- {os.path.basename(CFG_FILE)}
- {os.path.basename(ENV_FILE)}

""")

    # Print success message
    print("OK: IST-Zustand erfasst.")
    print(f"Ausgabeordner: {OUT_DIR}")
    print("Dateien:")
    try:
        for item in sorted(os.listdir(OUT_DIR)):
            print(f"- {item}")
    except FileNotFoundError:
        pass

if __name__ == "__main__":
    main()

File: javascript/collect_ist_gateway_b.js
#!/usr/bin/env node
// collect_ist_gateway_b.sh — portiert nach javascript
// Quelle: shell, OpenClaw@gateway2:scripts/collect_ist_gateway_b.sh
// Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

import fs from 'fs';
import os from 'os';
import path from 'path';
import { execSync } from 'child_process';

const BASE_DIR = path.join(os.homedir(), '.openclaw');
const OUT_DIR = path.join(BASE_DIR, 'workspace', 'vscode');
const NOW_UTC = new Date().toISOString().replace(/\.\d+Z$/, 'Z');
const NOW_LOCAL = new Date().toLocaleString('sv-SE') + ' ' + Intl.DateTimeFormat().resolvedOptions().timeZone;
const TS = new Date().toISOString().replace(/[-:]/g, '').replace('T', '-').substring(0, 15);

fs.mkdirSync(OUT_DIR, { recursive: true });

const IST_FILE = path.join(OUT_DIR, 'IST-ZUSTAND_GATEWAY-B_NODE7.md');
const INV_FILE = path.join(OUT_DIR, 'ARTEFAKT-INVENTAR_GATEWAY-B_NODE7.md');
const CFG_FILE = path.join(OUT_DIR, 'OPENCLAW-CONFIG-SNAPSHOT_GATEWAY-B_NODE7.md');
const ENV_FILE = path.join(OUT_DIR, 'ENV-STATUS_GATEWAY-B_NODE7.md');
const RUN_FILE = path.join(OUT_DIR, `RUN-${TS}.md`);

const OPENCLAW_JSON = path.join(BASE_DIR, 'openclaw.json');
const ENV_DOT = path.join(BASE_DIR, '.env');
const ENV_SYSTEMD = path.join(BASE_DIR, 'gateway.systemd.env');
const VSCODE_DIR = path.join(BASE_DIR, '.vscode');

let HOSTNAME_FQDN;
try {
  HOSTNAME_FQDN = os.hostname();
} catch (e) {
  HOSTNAME_FQDN = 'unbekannt';
}
const HOSTNAME_SHORT = os.hostname();
const ARCH = process.arch;
const KERNEL = os.release();
let OS_PRETTY = 'unbekannt';
try {
  const osRelease = fs.readFileSync('/etc/os-release', 'utf8');
  const match = osRelease.match(/^PRETTY_NAME=(.*)$/m);
  if (match) OS_PRETTY = match[1].replace(/"/g, '');
} catch (e) {}

const IPV4_ALL = Object.values(os.networkInterfaces())
  .flat()
  .filter(iface => !iface.internal && iface.family === 'IPv4')
  .map(iface => iface.address)
  .join(' ');

let PUBLIC_IP = '(nicht ermittelt)';
try {
  PUBLIC_IP = execSync('curl -4 -s --max-time 4 ifconfig.me', { encoding: 'utf8' }).trim();
} catch (e) {}

let TAILSCALE_IP = '(nicht ermittelt)';
try {
  TAILSCALE_IP = execSync('tailscale ip -4', { encoding: 'utf8' }).split('\n')[0].trim();
} catch (e) {}

let OPENCLAW_VER = '(nicht ermittelt)';
try {
  OPENCLAW_VER = execSync('openclaw --version', { encoding: 'utf8' }).trim();
} catch (e) {}

let NODE_VER = '(nicht ermittelt)';
try {
  NODE_VER = process.version;
} catch (e) {}

const istContent = `# IST-Zustand: Gateway B / Node 7

Stand (lokal): ${NOW_LOCAL}  
Stand (UTC): ${NOW_UTC}

## 1) Identität & System

- Gateway: **B**
- Node: **7**
- Hostname (short): \`${HOSTNAME_SHORT}\`
- Hostname (FQDN): \`${HOSTNAME_FQDN}\`
- Architektur: \`${ARCH}\`
- Kernel: \`${KERNEL}\`
- OS: \`${OS_PRETTY}\`
- IPv4 (lokal): \`${IPV4_ALL}\`
- Public IPv4: \`${PUBLIC_IP}\`
- Tailscale IPv4: \`${TAILSCALE_IP}\`
- OpenClaw Version: \`${OPENCLAW_VER}\`
- Node.js Version: \`${NODE_VER}\`

## 2) Arbeitsverzeichnisse

- Basis: \`${BASE_DIR}\`
- Funktionell VSCode: \`${VSCODE_DIR}\`
- Workspace Doku: \`${OUT_DIR}\`

## 3) Kernartefakte (Existenz)

- \`${OPENCLAW_JSON}\`: ${fs.existsSync(OPENCLAW_JSON) ? 'vorhanden' : 'fehlt'}
- \`${ENV_DOT}\`: ${fs.existsSync(ENV_DOT) ? 'vorhanden' : 'fehlt'}
- \`${ENV_SYSTEMD}\`: ${fs.existsSync(ENV_SYSTEMD) ? 'vorhanden' : 'fehlt'}
- \`${path.join(BASE_DIR, 'plugins/installs.json')}\`: ${fs.existsSync(path.join(BASE_DIR, 'plugins/installs.json')) ? 'vorhanden' : 'fehlt'}
- \`${path.join(BASE_DIR, 'plugin-skills')}\`: ${fs.existsSync(path.join(BASE_DIR, 'plugin-skills')) ? 'vorhanden' : 'fehlt'}

## 4) Hinweis

Diese Datei wird bei jedem Lauf neu geschrieben.
Zusätzlich wird ein Laufprotokoll als \`RUN-*.md\` erzeugt.`;

fs.writeFileSync(IST_FILE, istContent);

let invContent = `# Artefakt-Inventar: Gateway B / Node 7

Stand: ${NOW_LOCAL}

## Top-Level in ~/.openclaw

\`\`\`text
`;
try {
  invContent += fs.readdirSync(BASE_DIR).join('\n') + '\n';
} catch (e) {
  invContent += '(Fehler beim Lesen)\n';
}
invContent += '```\n\n## ~/.openclaw/.vscode\n\n```text\n';
if (fs.existsSync(VSCODE_DIR)) {
  try {
    const files = fs.readdirSync(VSCODE_DIR);
    files.forEach(file => {
      const stat = fs.statSync(path.join(VSCODE_DIR, file));
      const mode = stat.mode.toString(8).slice(-3);
      invContent += `${stat.isDirectory() ? 'd' : '-'}${mode} ${stat.size} ${file}\n`;
    });
  } catch (e) {
    invContent += '(Fehler beim Lesen)\n';
  }
} else {
  invContent += '(nicht vorhanden)\n';
}
invContent += '```\n\n## plugin-skills/\n\n```text\n';
if (fs.existsSync(path.join(BASE_DIR, 'plugin-skills'))) {
  try {
    invContent += fs.readdirSync(path.join(BASE_DIR, 'plugin-skills')).join('\n') + '\n';
  } catch (e) {
    invContent += '(Fehler beim Lesen)\n';
  }
} else {
  invContent += '(nicht vorhanden)\n';
}
invContent += '```\n\n## openclaw.json Backups\n\n```text\n';
try {
  const files = fs.readdirSync(BASE_DIR).filter(f => f.startsWith('openclaw.json.bak'));
  if (files.length > 0) {
    invContent += files.join('\n') + '\n';
  } else {
    invContent += '(keine gefunden)\n';
  }
} catch (e) {
  invContent += '(keine gefunden)\n';
}
invContent += '```';

fs.writeFileSync(INV_FILE, invContent);

let cfgContent = `# OpenClaw Config Snapshot: Gateway B / Node 7

Stand: ${NOW_LOCAL}

## Schlüsselpositionen (grep)

\`\`\`text
`;
if (fs.existsSync(OPENCLAW_JSON)) {
  try {
    const content = fs.readFileSync(OPENCLAW_JSON, 'utf8');
    const lines = content.split('\n');
    lines.forEach((line, index) => {
      if (line.match(/"gateway"|\"session\"|\"dmScope\"|\"auth\"|\"secrets\"|\"tools\"|\"plugins\"|\"profile\"|\"alsoAllow\"|\"denyCommands\"/)) {
        cfgContent += `${index + 1}: ${line}\n`;
      }
    });
  } catch (e) {
    cfgContent += 'Fehler beim Lesen\n';
  }
} else {
  cfgContent += 'openclaw.json fehlt\n';
}
cfgContent += '```\n\n## Ausschnitt gateway/session/auth (ungefiltert, betriebsnah)\n\n```json\n';
if (fs.existsSync(OPENCLAW_JSON)) {
  try {
    const content = fs.readFileSync(OPENCLAW_JSON, 'utf8');
    const lines = content.split('\n');
    const start = 579;
    const end = 779;
    for (let i = start; i <= Math.min(end, lines.length - 1); i++) {
      cfgContent += lines[i] + '\n';
    }
  } catch (e) {
    cfgContent += '{ "error": "Fehler beim Lesen" }\n';
  }
} else {
  cfgContent += '{ "error": "openclaw.json fehlt" }\n';
}
cfgContent += '```';

fs.writeFileSync(CFG_FILE, cfgContent);

let envContent = `# ENV-Status: Gateway B / Node 7

Stand: ${NOW_LOCAL}

## Dateien

\`\`\`text
`;
try {
  if (fs.existsSync(ENV_DOT)) {
    const stat = fs.statSync(ENV_DOT);
    envContent += `-rwx------ ${stat.size} ${path.basename(ENV_DOT)}\n`;
  }
  if (fs.existsSync(ENV_SYSTEMD)) {
    const stat = fs.statSync(ENV_SYSTEMD);
    envContent += `-rwx------ ${stat.size} ${path.basename(ENV_SYSTEMD)}\n`;
  }
} catch (e) {}
envContent += '```\n\n## .env (vollständig, ungefiltert)\n\n```dotenv\n';
if (fs.existsSync(ENV_DOT)) {
  try {
    envContent += fs.readFileSync(ENV_DOT, 'utf8') + '\n';
  } catch (e) {
    envContent += '# Fehler beim Lesen\n';
  }
} else {
  envContent += '# .env fehlt\n';
}
envContent += '```\n\n## gateway.systemd.env (vollständig, ungefiltert)\n\n```dotenv\n';
if (fs.existsSync(ENV_SYSTEMD)) {
  try {
    envContent += fs.readFileSync(ENV_SYSTEMD, 'utf8') + '\n';
  } catch (e) {
    envContent += '# Fehler beim Lesen\n';
  }
} else {
  envContent += '# gateway.systemd.env fehlt\n';
}
envContent += '```';

fs.writeFileSync(ENV_FILE, envContent);

const runContent = `# Laufprotokoll Gateway B / Node 7

- Zeit (lokal): ${NOW_LOCAL}
- Zeit (UTC): ${NOW_UTC}
- Script: ${process.argv[1]}

## Erzeugte Dateien

- ${path.basename(IST_FILE)}
- ${path.basename(INV_FILE)}
- ${path.basename(CFG_FILE)}
- ${path.basename(ENV_FILE)}
`;

fs.writeFileSync(RUN_FILE, runContent);

console.log('OK: IST-Zustand erfasst.');
console.log(`Ausgabeordner: ${OUT_DIR}`);
console.log('Dateien:');
const files = fs.readdirSync(OUT_DIR);
files.forEach(file => console.log(`- ${file}`));
